### PR TITLE
Add Option to run Incremental Export From Bigtable (Timestamp Filtering)

### PR DESCRIPTION
**Pull request by maheshgoyal15**
**Description:**

This pull request introduces a new feature to enable incremental data exports from Google Cloud Bigtable based on timestamp filtering. This addresses a specific customer requirement for efficiently extracting daily data subsets for transfer to a non-production Bigtable instance.

**Problem:**

The customer needs to export data from their production Bigtable database on a daily basis, filtering it based on a timestamp criterion. Currently, the existing export functionality lacks the ability to perform incremental exports based on timestamps, forcing them to perform full table scans or implement complex custom solutions.

**Solution:**

This PR implements the following changes to enable incremental exports based on timestamps:

* **Introduce Timestamp Filtering Mechanism:**
    * Adds two new command-line options: `--startTimestamp` and `--endTimestamp`.
    * These options allow users to specify a time range in UTC format (YYYY-MM-DDTHH:MM:SSZ) to filter the data during export.
    * The export process will then only retrieve and export rows that fall within the specified time range.
* **Leverage Bigtable Timestamp Filtering Capabilities:**
    * Utilizes Bigtable's `TimestampRangeFilter` to efficiently retrieve only the required data.
    * This ensures that only the data within the specified time range is scanned and exported, minimizing resource consumption and export time.
* **Implementation Details:**
    * The `BigtableToAvro.Options` interface has been extended to include `getStartTimestamp()` and `getEndTimestamp()` methods.
    * A `timestampConverter` function is implemented to convert the provided UTC timestamp strings into microseconds.
    * A `filterProvider` `ValueProvider` is used to construct the `RowFilter` based on the provided start and end timestamps.
    * The `BigtableIO.Read` operation now includes a `withRowFilter(filterProvider)` call to apply the timestamp filter.
    * Validation is added to make sure if one time stamp is provided, the other one must be provided as well.
* **Code Snippet (RowFilter Construction):**

```java
ValueProvider<RowFilter> filterProvider = new DualInputNestedValueProvider<>(
        options.getStartTimestamp(),
        options.getEndTimestamp(),
        (TranslatorInput<String, String> input) -> {
            String startTimestamp = input.getX();
            String endTimestamp = input.getY();

            boolean hasStart = startTimestamp != null && !startTimestamp.isEmpty();
            boolean hasEnd = endTimestamp != null && !endTimestamp.isEmpty();

            // Check if exactly one timestamp is provided (which is invalid)
            if ((hasStart && !hasEnd) || (!hasStart && hasEnd)) {
                throw new IllegalArgumentException(
                        "Both startTimestamp and endTimestamp must be provided together, or neither should be provided.");
            }

            // If neither timestamp is provided, return null (no filter)
            if (!hasStart && !hasEnd) {
                return null;
            }

            // Convert timestamps to microseconds
            Long startMicros = timestampConverter.apply(startTimestamp);
            Long endMicros = timestampConverter.apply(endTimestamp);

            // Build the timestamp filter
            com.google.cloud.bigtable.data.v2.models.Filters.TimestampRangeFilter filterBuilder =
                    FILTERS.timestamp().range();

            if (startMicros != null) {
                filterBuilder.startClosed(startMicros);
            }
            if (endMicros != null) {
                filterBuilder.endOpen(endMicros);
            }

            return filterBuilder.toProto();
        });

read = read.withRowFilter(filterProvider);

### CODE REVIEW

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java [hunk 1]
```diff
@@ -46,10 +49,12 @@
...
 import org.apache.beam.sdk.transforms.SerializableFunction;
 import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
 
 /**
- * Dataflow pipeline that exports data from a Cloud Bigtable table to Avro files in GCS. Currently,
- * filtering on Cloud Bigtable table is not supported.
+ * Dataflow pipeline that exports data from a Cloud Bigtable table to Avro files in GCS.
+ * 2/25 Add filtering rows based on timestamp.
```
Comment (ron-gal) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java:57]: 2/25?

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java [hunk 2]
```diff
@@ -142,6 +147,31 @@ public interface Options extends PipelineOptions {
 
     @SuppressWarnings("unused")
     void setBigtableAppProfileId(ValueProvider<String> appProfileId);
+    @TemplateParameter.Text(
+        order = 7,
+        groupName = "Source",
+        optional = true,
+        regexes = {"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\\.[0-9]+)?Z"},
+        description = "Start Timestamp in UTC Format (YYYY-MM-DDTHH:MM:SSZ) for exporting ",
```
Comment (ron-gal) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java:155]: something in this line seems broken. If not, please remove the whitespace after "Exporting".

File: v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java [hunk 3]
```diff
@@ -142,6 +147,31 @@ public interface Options extends PipelineOptions {
...
+    @SuppressWarnings("unused")
+    void setStartTimestamp(ValueProvider<String> startTimestamp);
+
+    @TemplateParameter.Text(
+        order = 8,
+        groupName = "Source",
+        optional = true,
+        regexes = {"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\\.[0-9]+)?Z"},
+        description = "End Timestamp in UTC Format (YYYY-MM-DDTHH:MM:SSZ)",
+        helpText = " Example UTC timestamp 2024-10-27T10:15:30.00Z"
```
Comment (ron-gal) [inline v1/src/main/java/com/google/cloud/teleport/bigtable/BigtableToAvro.java:169]: missing text in the beginning

**Review by ron-gal — COMMENTED** [review]
Please also update the integration test to use the filter, and maybe add unit tests to test edge cases like start without end etc

**Review by svetakvsundhar — CHANGES_REQUESTED** [review]
Thanks @maheshgoyal15 ! Could you add some e2e tests [here](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/095c8d4f812b26d93fa8dc90c441126905f02632/v1/src/test/java/com/google/cloud/teleport/bigtable/BigtableToAvroTest.java) to ensure this WAI?

**Comment by maheshgoyal15** [comment]
@ron-gal There were few delays but I am able to incorporate all the recommendation you provided. Can you take a look at this PR

**Comment by damccorm** [comment]
@ron-gal could you please take another look?

**Review by ron-gal — APPROVED** [review]

**Comment by svetakvsundhar** [comment]
@damccorm/ @liferoad : could we merge?
